### ISSUE TITLE
Provide option to use spdlog as a (static) library

### ISSUE THREAD

**Issue by gregoire-astruc** [gabime/spdlog#120]
For a reasonable sized project, the logger is very likely to be all over the place.

Rebuilding the whole project because an implementation detail changed would be a pain.

Could we have a compile-time option (eg `SPDLOG_LIBRARY` or `SPDLOG_HEADER_ONLY`) that would allow somebody to choose between header-only or precompiled?

I noticed cppformat uses a similar approach and I'd really appreciate that option.

I can provide that if you're up for it.

**Comment by gabime** [comment]
Sounds good but since its a header only by design it probably won't be very trivial.. Although i did keep separatarion of impl headers and  headers.

**Comment by rpopescu** [comment]
@gregoire-astruc 
But an implementation detail change will only happen if you change the local copy of the spdlog library yourself, or decide to update it from the external repository, and that is not something that would likely happen even once a day, right?
Therefore this will just add more complexity to the library without any real benefit in the most common case, right?

**Comment by gregoire-astruc** [comment]
I see your point,  but I'm not quite sure. For instance cppformat is used in the project as header only whilst it defaults to a static one: as a client you used an option provided by the library.

Any project using spdlog is likely to be some orders of magnitude bigger, even simply looking at files count, than spdlog itself.

Now I'd expect the logging system to be all over the place. Even if you wrap it in another class, it still gets transitively included. 

Using the header only approach, any change would trigger almost a rebuild all of the application. 

Using the library approach, you'd rebuild spdlog and relink only, which should be the common case (no public api change). 

cppformat uses this *added complexity* and it's nice, as a user, to have the choice.

**Comment by gregoire-astruc** [comment]
@gabime : I was on Linux Mint.

**Comment by rpopescu** [comment]
@gregoire-astruc Either I don't get it, or you don't see my point, because the key part in your argument is "any change would trigger almost a rebuild all of the application."

What change? When would changes happen to spdlog, part of a larger application?
I think the answer is once in a blue moon: when the larger application owner changes spdlog manually, or it updates it from an external source. These are rare events, therefore there is no pain caused by spurious rebuilds.

**Comment by gregoire-astruc** [comment]
I did not miss your point and I do agree.

However, as I mentioned with cppformat, as the library client you had a *choice* and you used it.

As rare as you update any library it's not uncommon, and I'd appreciate the option to compile most of it aside a larger project.

This also means I *can* update more frequently as long as the public headers don't change, since it will only require a relink (well there's more to it, but mostly).

**Comment by rpopescu** [comment]
I get it - you're focused on having the option part, I'm focused on the not-paying-for-what-I-don't-use part.
Those things may co-exist :)

**Comment by gregoire-astruc** [comment]
Exactly: I'm all for libraries which gives me viable options fitting my needs and expectations (thus my contribution).

As for the second part well:
* With a static library, any non-template function is compiled only once and could still very well be inlined by the linker, optimized and all.
* With the header only version every function will be recompiled for any compilation unit using it and duplicates will be merged at link-time. That's two to three times the work, and I'd appreciate to limit that.

**Comment by gabime** [comment]
Right, although pre-compiled headers solve this too, so maybe its not critical

**Comment by morpav** [comment]
@gabime I'm afraid that pre-compiled headers are not a working solution here. I have evaluated using spdlog for our project few days ago and here is my experience and some numbers for the following minimal code:

```
#include "spdlog/spdlog.h"
int main() {
  auto console = spdlog::stdout_logger_mt("console", true);
  console->info("Hello world!") ;
  return 0;
}
```
With Clang++ 3.9 (i5-4670K @ 3.40GHz, Linux), with and without a pre-compiled header containing only "spdlog/spdlog.h", trying "debug" and "release" builds, the compilation times were the following:

| |  no optimizations | -O3 |
| --- | --- | --- |
| direct `#include` | **1.45s** | **2.4s**  |
| using PCH | **0.95s** | **1.85s** |

So this example alone takes almost a second to compile even with PCH and no optimizations. When I want to use a logging library almost in all source files, then every single time I change anything in my source code I need to wait 1-2s more for each affected source file (depending on the setup of course). It is a serious penalty in development time.

Interestingly, when I try to compile only `int main() { return 0; }` with the PCH, I still get **0.75s** in debug mode. So the feature it is not cheap at all.

Besides the not so great compilation speed, pre-compiled headers have plenty of their own problems.

I would hope that with a proper knowledge of the library internals, it should not be so difficult to allow users to select a desired behavior. I think, the ideal solution from user's perspective would be this: 

Allow to define a flag like `SPDLOG_NO_HEADER_ONLY` to force the library to include only the necessary internal parts when included. Then, user could include the "spdlog/spdlog.h" into one selected source file like this:
```
#define SPDLOG_PLACE_INTERNALS
#include "spdlog/spdlog.h"
```
And the non-header-only parts of the library would be compiled there. So no static library would be needed, build process would by the same and without setting the flag, everything would work as usual. 

The beauty of such solution is that simply by changing a flag from your build system, you could toggle between "header only" and "library" options. E.g. fast compilation for development, best optimization for release builds. It would be sexy :)

**Comment by rpopescu** [comment]
Pavel,

While the experiment is worthwhile, the conclusion that you need to wait
1-2 seconds per file using spdlog is not sound.

If you want to measure that overhead, you should write a few more files
with functions calling spdlog's API, and incrementally compile a series of
binaries consisting of an increasing number of source files (start with 1
as you did, then 2 files etc.).
Then you measure the difference between the build times of these binaries
and can draw a better conclusion, because as such what you're measuring
with just 1 file involves other processes which are not cheap at all, in
particular linking.

Cheers,
Radu


On 21 May 2016 at 17:27, Pavel Moravec <notifications@github.com> wrote:

> @gabime <https://github.com/gabime> I'm afraid that pre-compiled headers
> are not a working solution here. I have evaluated using spdlog for our
> project few days ago and here is my experience and some numbers for the
> following minimal code:
>
> #include "spdlog/spdlog.h"
> int main() {
>   auto console = spdlog::stdout_logger_mt("console", true);
>   console->info("Hello world!") ;
>   return 0;
> }
>
> With Clang++ 3.9 (i5-4670K @ 3.40GHz, Linux), with and without a
> pre-compiled header containing only "spdlog/spdlog.h", trying "debug" and
> "release" builds, the compilation times were the following:
> no optimizations -O3
> direct #include *1.45s* *2.4s*
> using PCH *0.95s* *1.85s*
>
> So this example alone takes almost a second to compile even with PCH and
> no optimizations. When I want to use a logging library almost in all source
> files, then every single time I change anything in my source code I need to
> wait 1-2s more for each affected source file (depending on the setup of
> course). It is a serious penalty in development time.
>
> Interestingly, when I try to compile only int main() { return 0; } with
> the PCH, I still get *0.75s* in debug mode. So the feature it is not
> cheap at all.
>
> Besides the not so great compilation speed, pre-compiled headers have
> plenty of their own problems.
>
> I would hope that with a proper knowledge of the library internals, it
> should not be so difficult to allow users to select a desired behavior. I
> think, the ideal solution from user's perspective would be this:
>
> Allow to define a flag like SPDLOG_NO_HEADER_ONLY to force the library to
> include only the necessary internal parts when included. Then, user could
> include the "spdlog/spdlog.h" into one selected source file like this:
>
> #define SPDLOG_PLACE_INTERNALS
> #include "spdlog/spdlog.h"
>
> And the non-header-only parts of the library would be compiled there. So
> no static library would be needed, build process would by the same and
> without setting the flag, everything would work as usual.
>
> The beauty of such solution is that simply by changing a flag from your
> build system, you could toggle between "header only" and "library" options.
> E.g. fast compilation for development, best optimization for release
> builds. It would be sexy :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/gabime/spdlog/issues/120#issuecomment-220783778>
>



-- 
Radu-Adrian Popescu
Mobile +31-(0)611 00 45 20

**Comment by morpav** [comment]
@rpopescu Thanks for your inputs to the discussion.

For clarification, the presented numbers are my _compilation_ times, not complete build times. And even if you add link time, the compilation time vastly dominates in this scenario.

One can try to compile as many files as wanted but since the compilation is independent for each file the associated price will be paid for every single compilation unit which includes the library, even transitively.

I tried to add the spdlog header to one of our common header files in our project to measure the impact on compilation time for a large set of different files. I tried various scenarios, without actually using spdlog at all (only including the header) or with using it etc. The details are not really interesting here since these measurements are not repeatable in different environments. But the measurements were consistent with the provided numbers. So I wanted to present the simplest example to make the point as clear and repeatable as possible.

I re-ran all the easily repeatable tests on fresh system without any other intensive processes running. Here it is, you can draw your own conclusion.

### Basic spdlog usage

```
#include "spdlog/spdlog.h"
int main() {
  auto console = spdlog::stdout_logger_mt("console", true);
  console->info("Hello world!") ;
  return 0;
}
```
| |  no optimizations | -O3 |
| --- | --- | --- |
| direct `#include` | **1.38s** | **2.37s**  |
| using PCH | 0.87s | 1.84s |

### Only included, not used
```
#include "spdlog/spdlog.h"
int main() { return 0; }
```
| |  no optimizations | -O3 |
| --- | --- | --- |
| direct `#include` | **1.22s** | **1.62s**  |
| using PCH | 0.68s | 1.07s |

### No include, no usage, only in pre-compiled header
```
int main() { return 0; }
```
| |  no optimizations | -O3 |
| --- | --- | --- |
| trivial main only | 0.015s | 0.017s  |
| using PCH | **0.68s** | **1.07s** |

The library is great! But I'm sure it could be easily much more usable if the "header only" behavior could be altered.

**Comment by gabime** [comment]
@morpav If you have clear idea how to conveniently support both modes, a pull request will be most welcome

**Comment by morpav** [comment]
@gabime Yes, I understand. Since I never saw the internals, I can provide only the idea right now. 

For any header-only library, it should be enough to separate the current code into two parts and solve small issues with inlines.

Part 1: Declaration of smallest possible interface to the library. E.g. declaration of public non-templated functions, forward declarations of internal classes etc. Plus templates needed by a library users. This should be a thin as possible. 

Part 2: Definitions of items from block 1. Actual implementation of all non-interface stuff.

It's like separating header and implementation file with a focus on hiding as much as possible in the implementation file. Let's assume we have two flags:

 * SPDLOG_LIBRARY - global flag, toggled from build system
 * SPDLOG_PLACE_INTERNALS - A flag used in exactly one compilation unit when one wants to use non-header-only option (SPDLOG_LIBRARY).

Then a pseudo-code of spdlog/spdlog.h could be like this:
```
#if (!defined SPDLOG_LIBRARY)
#define SPDLOG_INLINE inline
#else
#define SPDLOG_INLINE
#endif

#include part 1.
    .. containing something like
    class SomeClass {
     public:
      SPDLOG_INLINE void SomeMember();
      ...
    };

#if defined SPDLOG_PLACE_INTERNALS || !defined SPDLOG_LIBRARY
   #include  part 2
#endif
```

The smaller the part 1 is the better. A key thing is that the part 2 can provide all the implementation when used as header-only library, so everything is visible to a compiler to generate the most optimized code. On the other hand, if one include only the first two parts, most of the code should be placed in the one implementation file and a linker will do its work. Declarations from part 1 are typically really cheap. One exception is heavy usage of templates, but one pay the price only when it is really used.

The SPDLOG_INLINE stuff solves issues with multiple symbol definitions and/or undefined references in one or the second usage.

I'm sorry that I cannot provide a pull request but maybe someone can read this and help more.

Pavel

**Comment by tnovotny** [comment]
you can also read on some strategies to support static, dynamic and header only here
https://vittorioromeo.info/index/blog/2016_cpp_library_configuration_api.html

**Comment by Timmmm** [comment]
I think this would be a nice option (maybe using [the same approach as Loguru](https://github.com/emilk/loguru#compiling)). As it is, #including the spdlog header everywhere significantly slows down my build. Pre-compiled headers might help but they are a pain to set up.

Without spdlog:
```
real	0m40.314s
user	2m13.539s
sys	0m6.525s
```

Just adding `#include <spdlog/spdlog.h>` to our logging header:
```
real	1m8.682s
user	3m57.938s
sys	0m8.981s
```

That's 40% slower just for including a header!

**Comment by gabime** [comment]
Help is wanted for this issue. I simply don't have the bandwidth to deal with it

**Comment by vwbij** [comment]
Is it maybe a option to use the DLL done by me? I like more to use a DLL, the only drawback I see is that variadic argument list are not supported by good old DLL binding.

**Comment by gabime** [comment]
> Is it maybe a option to use the DLL done by me

What about .so/.a in Unix?

**Comment by vwbij** [comment]
Clearly, I still sit on the island "windows". For me, this does not change, so my DLL works like a charme, and static librarys has some drawbacks. Ok, my DLL is just for people that wants to use it (only) under Windows...

**Comment by Timmmm** [comment]
@vwbij How did you get around the problem of requiring fmt in the headers. As far as I can see there is no way to use fmt's nice variadic template system, and still have fast compilation times.

Or did you just resort to using printf and put up with the pain of printing uint64 etc?

**Comment by vwbij** [comment]
Yes, this is a hard cutback if you create a DLL based on simple exported functions. No variadic temlates possible at all, just the old printf style possible. 

I come around this by inventing a object with a streaming interface in spdloglb_vararg.h. If you have created such a object or inherit from this class, you can just stream in the parameters BEFORE call the logging functions, like:

logSink << MyInt64 << MyInt32 << MyFloat << MyString << MyWideString;
logSink.Write(MyLogLevel,MyFormatString);

That is not nearly smart as the variadic templates, but it is typesafe and follows the same formating conventions like the variadic template system. That is the drawback of the pure export function system of the DLL, have not found a better way around until now.

**Comment by Timmmm** [comment]
Since the variadic templates are the main thing preventing easily turning this into a normal library, and those from `fmt`, I did some tests to try and find the source of the slow compilation. With 100 `.cpp` files with three logging calls in each I found the following compilation times:

| Header Contents | Compilation Time |
|------------------|-------------------|
| `#include <spdlog/spdlog.h>` | 62 seconds |
| `#define SPDLOG_FMT_EXTERNAL 1`<br>`#include <spdlog/spdlog.h>` | 57 seconds |
| `#include <string>`<br>`void log(std::string&& s);` | 10 seconds |
| `#include <fmt/format.h>`<br>`void log(std::string&& s);` | 14 seconds |
| See below | 15 seconds |

```
#include <fmt/format.h>

void log(std::string&& s);

template <typename... Args>
void log(const char* s, const Args&... args)
{
	log(fmt::format(s, args...));
}
```

This suggests to me that most of the compilation time is due to spdlog rather than fmt. I therefore worked around this by using the final code above, and then I can only `#include <spdlog/spdlog.h>` from one `.cpp` file where `log(s)` is implemented.

**Comment by rpopescu** [comment]
Would you be so kind as to put up your test on Github?
I'd like to see how much it takes to build this on my development machine.
The reason is that I have most of the code in header files only, from spdlog and boost to internal code, with a single cpp main file, and it's not a trivial amount of code, and it does do a fair bit of logging amongst other things. And still a debug build takes 10s, while an optimised build takes 15s.

**Comment by Timmmm** [comment]
@rpopescu Here is my compilation time benchmark: https://github.com/Timmmm/SpdlogBench

**Comment by gabime** [comment]
@Timmmm Thanks for this. This is very useful - a good starting point to understand how to improve the complition times.

**Comment by rpopescu** [comment]
@Timmmm I'm afraid I don't have ninja nor meson, and to be honest no desire to install them; a simple cmake or makefile would have sufficed, I believe. I'll see what I can do about that, thanks for taking the time to do this.

**Comment by Timmmm** [comment]
@rpopescu: If you skim the `meson.build` you can see it is pretty easy - it should be trivial to make a `CMakeLists.txt`. I wouldn't recommend using a standard Makefile though because GNU make is relatively slow and will skew the result. Better to use CMake's ninja generator.

The `CMakeLists.txt` should be something like this (not tested):

```
cmake_minimum_required(VERSION 3.6) # For sanity

project(LogTest)
set(CMAKE_CXX_STANDARD 11)

set(src
	"fmt/fmt/container.h"
	"fmt/fmt/format.cc"
	"fmt/fmt/format.h"
	"fmt/fmt/ostream.cc"
	"fmt/fmt/ostream.h"
	"fmt/fmt/posix.cc"
	"fmt/fmt/posix.h"
	"fmt/fmt/printf.cc"
	"fmt/fmt/printf.h"
	"fmt/fmt/string.h"
	"fmt/fmt/time.h"

	"LogInterface.cpp"
	"LogInterface.h"
	"main.cpp"
	"unit.cpp"

	# These are all copies of unit.cpp
	"unit_1.cpp"
	"unit_2.cpp"
	"unit_3.cpp"
	"unit_4.cpp"
	"unit_5.cpp"
	"unit_6.cpp"
	"unit_7.cpp"
	"unit_8.cpp"
	"unit_9.cpp"
	"unit_10.cpp"
	"unit_11.cpp"
	"unit_12.cpp"
	"unit_13.cpp"
	"unit_14.cpp"
	"unit_15.cpp"
	"unit_16.cpp"
	"unit_17.cpp"
	"unit_18.cpp"
	"unit_19.cpp"
	"unit_20.cpp"
	"unit_21.cpp"
	"unit_22.cpp"
	"unit_23.cpp"
	"unit_24.cpp"
	"unit_25.cpp"
	"unit_26.cpp"
	"unit_27.cpp"
	"unit_28.cpp"
	"unit_29.cpp"
	"unit_30.cpp"
	"unit_31.cpp"
	"unit_32.cpp"
	"unit_33.cpp"
	"unit_34.cpp"
	"unit_35.cpp"
	"unit_36.cpp"
	"unit_37.cpp"
	"unit_38.cpp"
	"unit_39.cpp"
	"unit_40.cpp"
	"unit_41.cpp"
	"unit_42.cpp"
	"unit_43.cpp"
	"unit_44.cpp"
	"unit_45.cpp"
	"unit_46.cpp"
	"unit_47.cpp"
	"unit_48.cpp"
	"unit_49.cpp"
	"unit_50.cpp"
	"unit_51.cpp"
	"unit_52.cpp"
	"unit_53.cpp"
	"unit_54.cpp"
	"unit_55.cpp"
	"unit_56.cpp"
	"unit_57.cpp"
	"unit_58.cpp"
	"unit_59.cpp"
	"unit_60.cpp"
	"unit_61.cpp"
	"unit_62.cpp"
	"unit_63.cpp"
	"unit_64.cpp"
	"unit_65.cpp"
	"unit_66.cpp"
	"unit_67.cpp"
	"unit_68.cpp"
	"unit_69.cpp"
	"unit_70.cpp"
	"unit_71.cpp"
	"unit_72.cpp"
	"unit_73.cpp"
	"unit_74.cpp"
	"unit_75.cpp"
	"unit_76.cpp"
	"unit_77.cpp"
	"unit_78.cpp"
	"unit_79.cpp"
	"unit_80.cpp"
	"unit_81.cpp"
	"unit_82.cpp"
	"unit_83.cpp"
	"unit_84.cpp"
	"unit_85.cpp"
	"unit_86.cpp"
	"unit_87.cpp"
	"unit_88.cpp"
	"unit_89.cpp"
	"unit_90.cpp"
	"unit_91.cpp"
	"unit_92.cpp"
	"unit_93.cpp"
	"unit_94.cpp"
	"unit_95.cpp"
	"unit_96.cpp"
	"unit_97.cpp"
	"unit_98.cpp"
	"unit_99.cpp"
	"unit_100.cpp"
)

include_directories("spdlog/include" "fmt")
add_definitions("-DVARIANT_${VARIANT}")
set(CMAKE_CXX_FLAGS "-Wno-unused-function")

add_executable("logtest" ${src})
```

And the `benchmark.sh` should be changed to something like this (also not tested):

```
#!/bin/bash

# Copy the `unit.cpp` file a load.
for n in {1..100}
do
	cp unit.cpp unit_$n.cpp
done

# Configure the builds.
for variant in A B C D E F
do
	mkdir build_$variant
	cd build_$variant
	cmake -DVARIANT=$variant -G"Ninja Makefiles" ..
	cd ..
done

# Time compilation
for variant in A B C D E F
do
	cd build_$variant
	time ninja
	cd ..
done
```

But really, if you are using Linux or OSX and already have Python 3 it is trivial to install meson, and it doesn't pull any crazy dependencies:

    python3 -m pip install meson

And if ninja is not available in your package repository, or is too old because your distro is stupid and thinks everyone loves out of date 'stable' software, then you can easily download binaries from their website:

```
wget https://github.com/ninja-build/ninja/releases/download/v1.7.2/ninja-linux.zip
unzip ninja-linux.zip
chmod +x ninja
sudo cp ninja /usr/local/bin/
```

Anyway feel free to try CMake/ninja, but I would have thought installing Meson is about 10 times quicker. Or... I put the benchmark results in the readme so you don't really need to run it at all. :-)

**Comment by rpopescu** [comment]
Actually I have FC24 on this machine and meson/ninja were available so that was easy enough :)
Here are my results.
I'm getting just under 2s for variant A, and 24s for variant D; I didn't do the others.
With -O3 I get 3s and 36s respectively.
This particular comparison is hardly fair - the code that is actually compiled and generated differs massively, as the A variant does almost nothing by comparison; the binary size is 11 times larger with spdlog without optimisations, and 28(!) times larger with -O3. I should think that in this case code parsing, generation and linking account for a large part of the difference.

**Comment by Timmmm** [comment]
Yeah to be clear I'm wasn't saying all the variants do the same thing. They were just interesting options to compare. A was meant to provide a reference point for how fast the compilation could be without any features.

**Comment by rpopescu** [comment]
Of course, I get it. What are your results? This is with GCC 6.3.1 btw, on Intel(R) Xeon(R) CPU E5-1680 v4 @ 3.40GHz, 32 GB RAM.

**Comment by Timmmm** [comment]
```
A
real    0m5.403s
user    0m27.527s
sys     0m3.558s

B
real    0m9.138s
user    0m47.499s
sys     0m4.686s

C
real    0m9.946s
user    0m48.817s
sys     0m4.882s

D
real    0m41.674s
user    3m48.771s
sys     0m13.925s

E
real    0m38.772s
user    3m32.701s
sys     0m13.462s
```

Spdlog adds almost exactly 2 seconds of user time per C++ file, compared to the plain function case (A).

This is on a 2016 Macbook Pro; 2.7 GHz i7, using Apple's LLVM 8.1.0.

**Comment by yisonPylkita** [comment]
Any progress on this feature request ?

**Comment by gabime** [comment]
Not really. Help is wanted

**Comment by yisonPylkita** [comment]
I like this library - clean interface and use of `fmt` library. I used it in a small projects before but now I need to replace a logger in a large project at my workplace and the only thing  stopping me from using `spdlog` is application build time with `spdlog`. I will make a R&D and check if building `spdlog` as shared/static library decreases compilation times.

**Comment by gabime** [comment]
You can try pre conpuled headers.. Should help

**Comment by yisonPylkita** [comment]
Maybe, but I will not follow that direction. There is something about precompiled headers that makes me feel uneasy.

I still have some time before deadline, so I will experiment with static/shared library approach.

I expect submitting a pull request in one week from now also with few *time to compile* benchmarks:
 - before changes
 - with clean `spdlog`
 - with clean `spdlog` and PCH
 - with `spdlog` as static/shared library
 - with `spdlog` as static/shared library and PCH

My codebase is something around 45k LOC

**Comment by Timmmm** [comment]
@yisonPylkita: Could you also do any of the cases in my benchmark? Especially the case where I just use `fmt`, in non-header-only mode (option C)?

**Comment by naveenpinto** [comment]
@vwbij i wanted to know if you have sources for your dll implementation using spd log, it will be a real help for my project, as i am stuck with the variadic interface which is not helping have DLL.

**Comment by YukinoHayakawa** [comment]
Another issue caused by the header only approach might be the symbol pollution caused by Windows headers. From time to time I have some identifiers unusable and it always tracks down to the Windows headers included by spdlog, deep in the including tree. I think I am going to take the wrapper approach by @Timmmm and don't expose spdlog header to my other files.

**Comment by DavidZemon** [comment]
Wow... long thread. I'm afraid I didn't read it all. But I found this thread because I wanted to know if anyone else had requested the same thing. We're starting to use spdlog at my company and each object file now takes a very long time to compile. If this gains any traction it's going to have a very painful effect on build times and I really don't want that. We all like the API and features a lot, but we really need to be able to compile this into a library (letting us choose shared vs. static, as is the norm for CMake projects, would be preferred).

**Comment by gabime** [comment]
Yes, this is a painful issue with spdlog. fwiw I plan to get back to this soon.

**Comment by DavidZemon** [comment]
Let me know if you need help with this from a build perspective. I have a number of years of CMake experience and can probably dedicate a bit of my work hours to helping make this happen.

**Comment by gabime** [comment]
Thanks, I will keep that in mind. My CMake skills are pretty basic.

**Comment by gabime** [comment]
(wip in branch static-lib)

**Comment by gabime** [comment]
@DavidZemon Could you review thw cmake in the static-lib branch? I think of modifying it according to https://steveire.wordpress.com/2016/08/09/opt-in-header-only-libraries-with-cmake/ 

provide 2 or 3 options to use the lib:
```
## uncomment only one of these!
# target_link_libraries(myexe 
#     spdlog::library_static)
# target_link_libraries(myexe
#     spdlog::library_iface)
target_link_libraries(myexe
     spdlog::library_srcs)
```

**Comment by DavidZemon** [comment]
Alright, PR opened as #1082

**Comment by gabime** [comment]
Implemented.

**Comment by DavidZemon** [comment]
I see the readme is recommending people use the static library already. Very cool. Does that also mean you will be releasing a new version soon?

**Comment by gabime** [comment]
@DavidZemon Yes I hope so. 
I  am already using the compiled version almost exclusively. It is great improvement  imo.
